pkg/ext-proc/backend/provider.go

```diff
@@ -16,127 +17,75 @@ const (
 	fetchMetricsTimeout = 5 * time.Second
 )
 
-func NewProvider(pmc PodMetricsClient, datastore *K8sDatastore) *Provider {
+func NewProvider(pmc PodMetricsClient, datastore Datastore) *Provider {
 	p := &Provider{
-		podMetrics: sync.Map{},
-		pmc:        pmc,
-		datastore:  datastore,
+		pmc:       pmc,
+		datastore: datastore,
 	}
 	return p
 }
 
 // Provider provides backend pods and information such as metrics.
 type Provider struct {
-	// key: Pod, value: *PodMetrics
-	podMetrics sync.Map
-	pmc        PodMetricsClient
-	datastore  *K8sDatastore
+	pmc       PodMetricsClient
+	datastore Datastore
 }
 
 type PodMetricsClient interface {
-	FetchMetrics(ctx context.Context, pod Pod, existing *PodMetrics) (*PodMetrics, error)
+	FetchMetrics(ctx context.Context, existing *PodMetrics) (*PodMetrics, error)
 }
 
-func (p *Provider) AllPodMetrics() []*PodMetrics {
-	res := []*PodMetrics{}
-	fn := func(k, v any) bool {
-		res = append(res, v.(*PodMetrics))
-		return true
-	}
-	p.podMetrics.Range(fn)
-	return res
-}
-
-func (p *Provider) UpdatePodMetrics(pod Pod, pm *PodMetrics) {
-	p.podMetrics.Store(pod, pm)
-}
-
-func (p *Provider) GetPodMetrics(pod Pod) (*PodMetrics, bool) {
-	val, ok := p.podMetrics.Load(pod)
-	if ok {
-		return val.(*PodMetrics), true
-	}
-	return nil, false
-}
-
-func (p *Provider) Init(logger logr.Logger, refreshPodsInterval, refreshMetricsInterval, refreshPrometheusMetricsInterval time.Duration) error {
-	p.refreshPodsOnce()
-
-	if err := p.refreshMetricsOnce(logger); err != nil {
-		logger.Error(err, "Failed to init metrics")
-	}
-
-	logger.Info("Initialized pods and metrics", "metrics", p.AllPodMetrics())
-
-	// periodically refresh pods
-	go func() {
-		for {
-			time.Sleep(refreshPodsInterval)
-			p.refreshPodsOnce()
-		}
-	}()
-
+func (p *Provider) Init(ctx context.Context, refreshMetricsInterval, refreshPrometheusMetricsInterval time.Duration) error {
 	// periodically refresh metrics
+	logger := log.FromContext(ctx)
 	go func() {
 		for {
-			time.Sleep(refreshMetricsInterval)
-			if err := p.refreshMetricsOnce(logger); err != nil {
-				logger.V(logutil.DEFAULT).Error(err, "Failed to refresh metrics")
+			select {
+			case <-ctx.Done():
+				logger.V(logutil.DEFAULT).Info("Shutting down metrics prober")
+				return
+			default:
+				time.Sleep(refreshMetricsInterval)
+				if err := p.refreshMetricsOnce(logger); err != nil {
+					logger.V(logutil.DEFAULT).Error(err, "Failed to refresh metrics")
+				}
 			}
 		}
 	}()
 
 	// Periodically flush prometheus metrics for inference pool
 	go func() {
 		for {
-			time.Sleep(refreshPrometheusMetricsInterval)
-			p.flushPrometheusMetricsOnce(logger)
+			select {
+			case <-ctx.Done():
```

Comment:
In integration tests, those routines were leaking across test cases, causing a lot of problems (a provider from a previous test case overriding state of a different test case). With this, we are forcing them to shutdown with the manager.